File: frontend/lib/post-caption.ts
import type { CreativeBlueprintResponse, StructuredTextPayload } from "@/lib/api/contracts";

function normalizeTags(hashtags: string[] | undefined): string[] {
  return (hashtags || [])
    .map((tag) => String(tag || "").trim())
    .filter(Boolean)
    .map((tag) => (tag.startsWith("#") ? tag : `#${tag.replace(/^#+/, "")}`));
}

function buildFallbackCaption({
  platform,
  blueprint,
  generatedPayload,
}: {
  platform: string;
  blueprint?: CreativeBlueprintResponse | null;
  generatedPayload?: StructuredTextPayload | null;
}): string {
  const hook = blueprint?.hook?.trim() || "";
  const headline = generatedPayload?.headline?.trim() || blueprint?.headline?.trim() || "";
  const body = generatedPayload?.body?.trim() || blueprint?.body?.trim() || "";
  const supporting = blueprint?.supporting_line?.trim() || "";
  const cta = generatedPayload?.cta?.trim() || blueprint?.cta?.trim() || "";
  const storyFlow = (blueprint?.story_flow || []).map((line) => String(line || "").trim()).filter(Boolean);
  const proofPoints = (blueprint?.proof_points || []).map((line) => String(line || "").trim()).filter(Boolean);
  const hashtags = normalizeTags(
    generatedPayload?.hashtags?.length ? generatedPayload.hashtags : blueprint?.hashtags,
  );

  const normalized = platform === "x" ? "twitter" : platform;
  const paragraphs: string[] = [];
  const lead = hook || headline;
  if (lead) paragraphs.push(lead);

  for (const line of storyFlow.slice(0, 5)) {
    if (!paragraphs.includes(line)) paragraphs.push(line);
  }
  if (supporting && !paragraphs.includes(supporting)) paragraphs.push(supporting);

  if (body) {
    const chunks = body.includes("\n\n") ? body.split("\n\n").map((c) => c.trim()).filter(Boolean) : [body];
    for (const chunk of chunks) {
      if (!paragraphs.includes(chunk)) {
        paragraphs.push(normalized === "twitter" && chunk.length > 220 ? `${chunk.slice(0, 217).trim()}…` : chunk);
      }
    }
  }

  for (const point of proofPoints.slice(0, 2)) {
    if (!paragraphs.includes(point)) paragraphs.push(point);
  }

  if (cta && !paragraphs.some((p) => p.toLowerCase().includes(cta.toLowerCase()))) {
    paragraphs.push(cta);
  }

  if (!paragraphs.some((p) => p.includes("?"))) {
    const topic = headline || hook || "this";
    paragraphs.push(`What stands out to you about ${topic.replace(/\?+$/, "")}?`);
  }

  if (hashtags.length) {
    const limit = normalized === "linkedin" ? 5 : normalized === "instagram" ? 8 : 3;
    paragraphs.push(hashtags.slice(0, limit).join(" "));
  }

  return paragraphs.join("\n\n").trim();
}

export function buildPostCaption({
  platform,
  blueprint,
  generatedPayload,
}: {
  platform: string;
  blueprint?: CreativeBlueprintResponse | null;
  generatedPayload?: StructuredTextPayload | null;
}): string {
  const fromBlueprint = String(blueprint?.post_caption || "").trim();
  if (fromBlueprint) {
    return fromBlueprint;
  }
  return buildFallbackCaption({ platform, blueprint, generatedPayload });
}

File: app/services/post_caption.py
from __future__ import annotations

from app.graph.models.layer7c_models import CreativeBlueprint


def build_post_caption_from_blueprint(blueprint: CreativeBlueprint, platform: str = "instagram") -> str:
    """Build a platform-ready social caption when the LLM omitted post_caption."""
    if str(blueprint.post_caption or "").strip():
        return str(blueprint.post_caption).strip()

    platform_key = (platform or "instagram").strip().lower()
    if platform_key == "x":
        platform_key = "twitter"

    paragraphs: list[str] = []
    hook = str(blueprint.hook or "").strip()
    headline = str(blueprint.headline or blueprint.title or "").strip()
    body = str(blueprint.body or "").strip()
    supporting = str(blueprint.supporting_line or "").strip()

    lead = hook or headline
    if lead:
        paragraphs.append(lead)

    for line in list(blueprint.story_flow or [])[:5]:
        text = str(line or "").strip()
        if text and text not in paragraphs:
            paragraphs.append(text)

    if supporting and supporting not in paragraphs:
        paragraphs.append(supporting)

    if body:
        if platform_key == "twitter" and len(body) > 220:
            paragraphs.append(body[:217].rstrip() + "…")
        else:
            for chunk in [p.strip() for p in body.split("\n\n") if p.strip()]:
                if chunk not in paragraphs:
                    paragraphs.append(chunk)

    for point in list(blueprint.proof_points or [])[:2]:
        text = str(point or "").strip()
        if text and text not in paragraphs:
            paragraphs.append(text)

    cta = str(blueprint.cta or "").strip()
    if cta and not any(cta.casefold() in p.casefold() for p in paragraphs):
        if platform_key == "linkedin":
            paragraphs.append(cta)
        elif "?" not in cta:
            paragraphs.append(cta)

    if not any("?" in p for p in paragraphs):
        topic = headline or hook or "this"
        paragraphs.append(f"What stands out to you about {topic.rstrip('?')}?")

    tags = [
        str(tag).strip()
        for tag in (blueprint.hashtags or [])
        if str(tag).strip()
    ]
    tags = [t if t.startswith("#") else f"#{t.lstrip('#')}" for t in tags]
    if tags:
        limit = 5 if platform_key == "linkedin" else 8 if platform_key == "instagram" else 3
        paragraphs.append(" ".join(tags[:limit]))

    return "\n\n".join(paragraphs).strip()
